### ISSUE TITLE
Variable sized input and target tensors

### ISSUE THREAD

**Issue by noa** [twitter/torch-autograd#19]
In some cases it's useful to have variable sized input and target tensors and batch sizes, e.g. RNNs. This was reasonably efficient before the Nov 16 overhaul but I'm now seeing extreme slowdowns after the overhaul. Any plans to introduce a flag to switch between the pre- and post-overhaul behaviors, or suggestions on how to improve performance? I assume that moving to fixed BPTT dimension and batch size would address the speed issue (to match the provided LM example), but maybe there's another way to improve performance without changing the data?

**Comment by clementfarabet** [comment]
@noa , this package was actually designed specifically for this reason :-) – i.e. to handle variable input lengths and dynamic graphs at runtime. In this new version, what happens is that you pay an extra cost each time a new input config is seen, but then the resulting graph is cached, and re-used in the future. 

We use this current version to train LSTMs on captions / text, where the text length varies form 5 words to 60 words, which results in ~ 50 cached tapes; once you've seen them all once, everything is cached, and speed is maximal.

Does your application require a much larger nb of input configurations (which might not even been seen multiple times)? If so then you might not be benefiting from the caching. Would appreciate more details on this use case. We're planning on optimizing the cache-free version as well, so that deactivating the cache entirely gives you decent performance, in the case where you have a huge nb of input configs.

**Comment by noa** [comment]
Yep, I'm also working with text. It looks like for my data the caching penalty is being paid well into an epoch of training:

10 batches = 56s per batch (!!)
100 batches = 40s per batch
200 batches = 3s per batch
300 batches = 1s per batch
...

Pre-Nov 16 the timing was always well below 1s.

If an "input config" as far as caching is concerned applies to all dimensions of the input (e.g. both batch size and BPTT), then this might explain some of the big slowdown for me post-Nov 16 since most batches will have different dimensions.

The other part of the problem seems to be the time required to cache individual parts of the graph -- 1min seems excessive? It occurred to me that CUDA might be at fault here. All my inputs live on the device as CudaTensors, which might mean that autograd is allocating a bunch of device memory whenever it caches a graph? The other contributing factor might be the length of some of my inputs, sometimes in excess of 100.

Other environment details:
- luajit
- latest torch (trepl, cutorch, nn, ...)

**Comment by soumith** [comment]
@clementfarabet maybe there could be a disk cache that is preserved across sessions, for the case like @noa ?

**Comment by clementfarabet** [comment]
@soumith interesting idea. @luketwitter : the disk cache idea is potentially interesting, to avoid warm up overheads.

But in this case, I'm afraid the nb of potential tapes might just be too high. @luketwitter maybe we just need to work on making the tape generation faster as well, to avoid this initial overhead, or maybe for a case like that the completely uncached mode is just the best. 

@noa curious about one thing: are you modeling full text paragraphs? It's usually common to do truncated backprop in this case – see example here: https://github.com/twitter/torch-autograd/blob/master/examples/train-penn-lstm.lua#L93 where the LSTM state is passed through, but the bprop truncated. But not a general fix for sure.

**Comment by clementfarabet** [comment]
@noa – @luketwitter just brought up something: are you using `d(f)(...)` in your inner loop? You now need to call `df = d(f)` once, and then call `df(...)` in your inner loop, to benefit from caching.

**Comment by noa** [comment]
> @noa curious about one thing: are you modeling full text paragraphs? It's usually common to do truncated backprop in this case – see example here: https://github.com/twitter/torch-autograd/blob/master/examples/train-penn-lstm.lua#L93 where the LSTM state is passed through, but the bprop truncated. But not a general fix for sure.

Long character/symbol sequences actually. Unfortunately, truncated BPTT is a little tricky in my case since I'm using a weird aggregate loss criterion. I can probably get it to work with some combination of truncated BPTT and padding with 0s, but this isn't very flexible.

> @luketwitter just brought up something: are you using d(f)(...) in your inner loop? You now need to call df = d(f) once, and then call df(...) in your inner loop, to benefit from caching.

Nope, I switched to the new convention of calling df = d(f) once prior to actually using the function. I was hoping it might be this easy!

**Comment by luketwitter** [comment]
@noa Working on a fast un-compiled mode for you (previous behavior), shouldn't take too long. I would love to see your example though, to make sure you can't benefit from the compilation.

**Comment by noa** [comment]
@luketwitter Awesome, thanks. I'll see about constructing a self-contained minimal example for you in the meantime.

**Comment by noa** [comment]
Check out: https://github.com/noa/autograd_test

I tried it with the current revision and with PR #22, which seems a bit faster. Not sure if it's back to pre-Nov 16 speed yet, will need to do some more testing.

**Comment by luketwitter** [comment]
Thanks - super helpful. There's a dumb bottleneck when generating the code for the second iteration, as it tries to map the set of tensors it needs to the set the first iteration needed, to maximize reuse. Takes 60 seconds, with your 16000-long tape of instructions :)

**Comment by luketwitter** [comment]
Fixed - 60s -> 4-5s per step. 

After ~100 steps, it actually starts getting some code cache hits, and the average time per step starts going down, so it's worth using the optimizer after all. Currently 1.2s per step at step 300.

For comparison, the direct/unoptimized version takes 3-4s per step.

**Comment by noa** [comment]
@luketwitter Great, #23 seems to have done the trick. With CUDA I'm down to ~300ms per step past 100 batches or so, and I'm not seeing the big 50s+ delays in early steps either. Thanks!

**Comment by alexbw** [comment]
Closing, fixed with #23
